res/controllers/Denon-MC7000-scripts.js

```diff
@@ -816,25 +849,78 @@ MC7000.sortLibrary = function(channel, control, value) {
+            midi.sendShortMsg(0xB0 + deckNumber - 1, 0x1F, vuLevelOutValue);
+            MC7000.prevVuLevel[deckNumber - 1] = vuThreshold;
+        }
+    } else { // if not playing set zero, otherwise 1st LED stays active when < threashold
+        midi.sendShortMsg(0xB0 + deckNumber - 1, 0x1F, 0x00);
+    }
 };
 
-/* LEDs around Jog wheel */
-MC7000.JogLed = function(value, group) {
+// Spinning Platter LEDs & Slicer Loop PAD LEDs as beat counter
+// pulled together for the calculation to be done only once and then
+// send LED signals to Jog or PAD LEDs when needed.
+MC7000.TrackPositionLEDs = function(value, group) {
     // do nothing before track starts
     if (value === 0) {
         return;
     }
+    // lets define some variables first
     var deckNumber = script.deckFromGroup(group),
-        trackDuration = engine.getValue(group, "duration"),
-        position = value * trackDuration / 60 * MC7000.scratchParams.recordSpeed,
-        // LED ring contains 48 segments with each LED activated by the next even number
-        jogLedOutValue = 48 * 2,
-        activeLED = MC7000.isVinylMode[deckNumber - 1] ? Math.round(position * jogLedOutValue) % jogLedOutValue : value * jogLedOutValue;
+        trackDuration = engine.getValue(group, "duration"), // in seconds
+        beatLength = engine.getValue(group, "file_bpm") / 60, // in Beats Per Seconds
+        cuePosition = engine.getValue(group, "cue_point") / engine.getValue(group, "track_samplerate") / 2, // in seconds
+        playPosition = value * trackDuration, // in seconds
+        jogLEDPosition = playPosition / 60 * MC7000.scratchParams.recordSpeed,
+        jogLEDNumber = 48, // LED ring contains 48 segments each triggered by the next even Midi value
+        // check for Vinyl Mode and decide to spin the Jog LED or show play position
+        activeJogLED = MC7000.isVinylMode[deckNumber - 1] ? Math.round(jogLEDPosition * jogLEDNumber) % jogLEDNumber : Math.round(value * jogLEDNumber),
+        // count the beats (1 to 8) behind the CUE point
+        beatCountLED = (Math.floor((playPosition - cuePosition) * beatLength) % 8); //calculate PAD LED position
+
+    // Todo: check for playposition < (trackduration - warning length) for sending position signals
+
+    // check if a Jog LED has changed and if so then send the signal to the next Jog LED
+    if (MC7000.prevJogLED[deckNumber - 1] !== activeJogLED) {
+        midi.sendShortMsg(0x90 + deckNumber - 1, 0x06, activeJogLED * 2); // only each 2nd midi signal triggers the next LED
+        MC7000.prevJogLED[deckNumber - 1] = activeJogLED;
+    }
+
+    // Todo: else blink the platter LEDs
 
-    midi.sendShortMsg(0x90 + deckNumber - 1, 0x06, activeLED);
+    // check if Slicer mode is active and illuminate PAD LEDs counting with the beat while playing
+    if (MC7000.experimental) {
+        if (MC7000.PADModeSlicer[deckNumber] && engine.getValue(group, "play_indicator")) {
```

Comment:
style-wise, I personally prefer flat functions using early returns, because I don't have to hold all outer expressions in my head while reading it. This obviously only works for if-statements without an else case, but there are a lot of those nested here so an early return would make sense IMO.
```suggestion
    if (!MC7000.experimental) {
        return;
    }
    if (MC7000.PADModeSlicer[deckNumber] && engine.getValue(group, "play_indicator")) {
        etc...
```
you don't have to change it if you disagree though.

Reply:
Well - that is a good point - thinking about that ....

Reply:
I changed it as I also like this way. Thanks